File: ideal-core/src/main/java/cn/idealframework2/operation/block/Operations.java
package cn.idealframework2.operation.block;

import cn.idealframework2.operation.OperationLog;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * @author 宋志宗 on 2022/12/7
 */
public final class Operations {
  private static final ThreadLocal<OperationLog> THREAD_LOCAL = new ThreadLocal<>();

  public static void set(@Nonnull OperationLog operationLog) {
    THREAD_LOCAL.set(operationLog);
  }

  @Nullable
  public static OperationLog get() {
    return THREAD_LOCAL.get();
  }

  public static void remove() {
    THREAD_LOCAL.remove();
  }

  private Operations() {
  }
}

File: ideal-core/src/main/kotlin/cn/idealframework2/operation/coroutine/Operations.kt
package cn.idealframework2.operation.coroutine

import cn.idealframework2.operation.OperationLog
import cn.idealframework2.operation.reactive.Operations
import kotlinx.coroutines.reactor.awaitSingle

/**
 * @author 宋志宗 on 2022/9/24
 */
@Suppress("unused")
object Operations {

  suspend fun awaitOperationLog(): OperationLog? {
    return Operations.current().awaitSingle().orElse(null)
  }

  suspend inline fun message(block: () -> String) {
    val operationLog = awaitOperationLog() ?: return
    operationLog.message = block.invoke()
  }

  suspend inline fun details(block: () -> String) {
    val operationLog = awaitOperationLog() ?: return
    operationLog.details = block.invoke()
  }

  suspend inline fun failure(block: () -> String) {
    val operationLog = awaitOperationLog() ?: return
    operationLog.isSuccess = false
    operationLog.message = block.invoke()
  }

  /** 设置变更前的信息 */
  suspend inline fun before(block: () -> String) {
    val operationLog = awaitOperationLog() ?: return
    operationLog.before = block.invoke()
  }

  /** 设置变更后的信息 */
  suspend inline fun after(block: () -> String) {
    val operationLog = awaitOperationLog() ?: return
    operationLog.after = block.invoke()
  }

  suspend inline fun log(block: (OperationLog) -> Unit) {
    val operationLog = awaitOperationLog() ?: return
    block.invoke(operationLog)
  }
}
